### DISCUSSION TITLE
Hidden context element  has height=0 when calling computePosition

### DISCUSSION THREAD

**Discussion by datrinh** [floating-ui/floating-ui#1525] [Q&A]
Hi,

I created a rather normal popover element using floating-ui. The issues arises when I want to use middlewares like `shift()`. I pass it the context element, but at the time of execution, the height of the DOM element is 0, because it's hidden (using vue's `v-if`).

I solved this rather dirty by wrapping the `update()` function into a `setTimeout`, so that the execution waits a tick, so that the height is calculated when update is triggered. Is there a more elegant (built-in) solution for this?

```js
  const update = async () => {
    if (contextEl.value) {
      const { x: posX, y: posY } = await computePosition(virtualElObj, contextEl.value.$el, {
        placement: 'bottom-start',
        middleware: [offset(-16), shift({ padding: 8, crossAxis: true })],
        ...floatUiOptions,
      })
      x.value = posX
      y.value = posY
    }
  }

  watch(boundingRect, () => {
    setTimeout(() => {
      // Wait a tick to compute height of contextEl
      update()
    }, 0)
  })
```

**Comment by atomiks** [comment]
This seems like a Vue-specific question. There's a "nextTick" in Vue iirc but not familiar enough with it.

**Reply by datrinh** [reply]
Is it really vue-specific tho? I could toggle visibility with vanilla JS/CSS too, and the behavior should be the same in theory? The context element has height of 0, since the element is hidden/empty. That timing issue should be there regardless of the underlying framework?

**Reply by atomiks** [reply]
If the element has a height of 0 when you call `computePosition()`, then that's the behavior of the framework. `computePosition()` is just a function call, it needs to have the element on the DOM to measure before doing any computations.

So it's like this in vanilla JS:

```js
document.body.appendChild(floatingElement);
updatePosition(floatingElement);
```

Whereas the situation you're seeing is:

```js
// the floating element can't be measured yet
updatePosition(floatingElement);
document.body.appendChild(floatingElement);
```

So it doesn't have anything to do with the lib

**Reply by datrinh** [reply]
> Whereas the situation you're seeing is:
> // the floating element can't be measured yet
> updatePosition(floatingElement);
> document.body.appendChild(floatingElement);

No, that's not right. The element is appended to the DOM way before, but it simply has a height of 0 until it should be visible. I don't see how it has anything to do with Vue. If I wanted to implement a tooltip (vanilla JS), that scales along the y-axis when active, I would have the same issue. To make my point clearer: If I use opacity instead of height, it works as expected, because the browser knows the height even when it's hidden. In both cases it's already existing on the DOM.

> So it doesn't have anything to do with the lib

So the gist is that there is no built-in solution for this timing issue. It's not a big deal really, just wanted to make sure that I don't miss anything.

**Reply by atomiks** [reply]
Well the logic is still the same: the mutation to make the height `auto` (or not 0 or whatever) is happening *after* the call to `computePosition()`. 

Since it's just a function call, it doesn't know anything about the element you're passing to it. That's what Vue's job is, you just need to wait for the element to actually have a height so it can be measured. Vue is doing the mutation after Floating UI does its work; so yeah no built-in way because it doesn't really make sense to. It's like needing a function `Math.round()` to "know" when to perform the right calculation, when it's agnostic/pure

**Reply by datrinh** [reply]
No complains about it not being handled by the lib :) Just asking about best practice here, since it seems like a common use case to me (regardless of UI framework). Not sure why you keep insisting on that it has anything to do with Vue though. The same applies to any vanilla JS setup or React for that matter. Thanks for the clarification in any case :)

**Reply by atomiks** [reply]
Well you're calling update inside `watch`, which is a Vue specific API. Why is `watch` not being called at the right time? In React the node's measurements will be correctly available inside `useLayoutEffect`

**Reply by datrinh** [reply]
Well, just because React provides a specialized hook that solves this issue, doesn't make it a Vue-specific issue - nor svelte-specific or vanillaJS-specific :) It's a general timing issue. Maybe it's obvious for most, but it might be beneficial to mention this pitfall in the documentation? Maybe I'm alone though.

**Reply by atomiks** [reply]
I don't know what `watch` does, I'd just assume that the callback Vue is calling inside there should be happening after the height style has been applied. So yes, while it's a timing issue, the key here is that the timing of the callback invoked by `watch` is wrong for some reason. Maybe not Vue's fault in particular, but the way it's being used _with_ Vue seems to be wrong? 

For docs we could mention that the floating element must exist on the DOM with styles applied for it to be measured when `computePosition()` is called.

**Reply by datrinh** [reply]
`watch` is effectively React's `watchEffect`. That's why I assumed that with `watchEffect` one might have the same result, but I didn't test it. In my specific case, I created a context menu on right click and `watch` listens to changes to the position of the pointer. The underlying issue might be that the `open` state and the `position` state are not tightly coupled, hence `open` might fire after `position`  - but that's just an educated guess. Maybe I'll refactor things to test it.
